src/VbsEnclaveSDK/src/veil_host_lib/hello_winrt_kcm.vtl0.cpp

```diff
@@ -0,0 +1,340 @@
+
+    // Extract the SID from the TOKEN_USER structure
+    PSID userSid = tokenUser->User.Sid;
+
+    // Convert SID to string
+    wil::unique_hlocal_ptr<WCHAR[]> userSidString;
+    THROW_LAST_ERROR_IF(!ConvertSidToStringSidW(userSid, wil::out_param_ptr<LPWSTR*>(userSidString)));
+
+    // Create the formatted key name
+    std::wstring result = std::format(c_formatString, userSidString.get(), userSidString.get(), name);
+
+    return result;
+}
+
+winrt::hstring GetAlgorithm(uintptr_t ecdhAlgorithm)
+{
+    using namespace winrt::Windows::Security::Cryptography::Certificates;
+    if (reinterpret_cast<BCRYPT_ALG_HANDLE>(ecdhAlgorithm) == BCRYPT_ECDH_P384_ALG_HANDLE)
+    {
+        return KeyAlgorithmNames::Ecdh384();
+    }
+    else if (reinterpret_cast<BCRYPT_ALG_HANDLE>(ecdhAlgorithm) == BCRYPT_ECDH_P256_ALG_HANDLE)
+    {
+        return KeyAlgorithmNames::Ecdh256();
+    }
+    THROW_HR(E_INVALIDARG);
+}
+
+// Helper function to convert veil_abi::Types::keyCredentialCacheConfig to KeyCredentialCacheConfiguration
+KeyCredentialCacheConfiguration ConvertCacheConfig(const veil_abi::Types::keyCredentialCacheConfig& cacheConfig)
+{
+    // Map cacheOption to KeyCredentialCacheOption
+    KeyCredentialCacheOption cacheOption;
+    switch (cacheConfig.cacheOption)
+    {
+        case 0:
+            cacheOption = KeyCredentialCacheOption::NoCache;
+            break;
+        case 1:
+            cacheOption = KeyCredentialCacheOption::CacheWhenUnlocked;
+            break;
+        default:
+            cacheOption = KeyCredentialCacheOption::NoCache; // Default fallback
+            break;
+    }
+
+    // Convert timeout from seconds to TimeSpan (100-nanosecond units)
+    winrt::Windows::Foundation::TimeSpan timeout{ static_cast<int64_t>(cacheConfig.cacheTimeoutInSeconds) * 10000000LL };
+
+    // Use WinRT's get_activation_factory for a more modern approach
+    auto factory = winrt::get_activation_factory<winrt::Windows::Security::Credentials::KeyCredentialCacheConfiguration, 
+                                                 winrt::Windows::Security::Credentials::IKeyCredentialCacheConfigurationFactory>();
+
+    return factory.CreateInstance(cacheOption, timeout, cacheConfig.cacheUsageCount);
+}
+
+veil_abi::Types::credentialAndSessionInfo veil_abi::Untrusted::Implementation::userboundkey_establish_session_for_create(
+    uintptr_t enclave,
+    const std::wstring& key_name,
+    uintptr_t ecdh_protocol,
+    const std::wstring& message,
+    uintptr_t window_id,
+    const veil_abi::Types::keyCredentialCacheConfig& cache_config,
+    uint32_t key_credential_creation_option)
+{
+    std::wcout << L"Inside userboundkey_establish_session_for_create_callback"<< std::endl;
+    auto algorithm = GetAlgorithm(ecdh_protocol);
+
+    // Convert the cacheConfig parameter to KeyCredentialCacheConfiguration
+    auto cacheConfiguration = ConvertCacheConfig(cache_config);
+
+    auto sessionInfo = std::make_shared<uintptr_t>(0);
```

Comment:
>uintptr_t [](http://example.com/codeflow?start=40&length=9)

memory: Why can't this just be a (vtl0-equivalent) of unique_sessionhandle?  Doesn't that solve all our memory issues w.r.t. leaking this session upon failure?